### PR TITLE
DAOS-16721 dtx: handle potential DTX ID reusing trouble - b26

### PR DESCRIPTION
**Pull request by Nasf-Fan**
The patch contains the following improvements:

1. When VOS level logic returns -DER_TX_RESATRT, the object level RPC handler should set 'RESEND' flag then restart the transaction with newer epoch. Because dtx_abort() logic cannot guarantee all former prepared DTX entries (on all related participants) can be aborted, especially if the former one failed for some network trouble, that may cause restarted transaction hit -DER_TX_ID_REUSED unexpectedly.

2. Compare the epoch for DTX entries with the same transaction ID for distinguishing potential reused TX ID more accurately.

3. Add DTX entry into DTX CoS cache if cannot commit it synchronously. Then subsequent batched commit logic can handle it.

4. If server complains suspected TX ID reusing, then reports -EIO to related application instead of assertion on client.

5. Control DTX related warning message frequency to avoid log flood.

6. Collect more information when generate some error/warning message.

Allow-unstable-test: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by liuxuezhao — DISMISSED** [review]

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -4834,11 +4834,14 @@ obj_comp_cb(tse_task_t *task, void *data)
 		D_ASSERT(daos_handle_is_inval(obj_auxi->th));
 		D_ASSERT(obj_is_modification_opc(obj_auxi->opc));
 
-		if (task->dt_result == -DER_TX_ID_REUSED && obj_auxi->retry_cnt != 0)
-			/* XXX: it is must because miss to set "RESEND" flag, that is bug. */
-			D_ASSERTF(0,
-				  "Miss 'RESEND' flag (%x) when resend the RPC for task %p: %u\n",
-				  obj_auxi->flags, task, obj_auxi->retry_cnt);
+		if (task->dt_result == -DER_TX_ID_REUSED && obj_auxi->retry_cnt != 0) {
+			D_ERROR("Be complained as TX ID reused for unknown reason, "
```
Comment (mchaarawi) [inline src/object/cli_obj.c:4838]: please remove the first part "Be complained as", as that is not a valid english phrase.

**Comment by phender** [comment]
The https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15409/13 build failed to run the Functional Hardware stages due to provisioning errors.  I've manually kicked off https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15409/14/ to just run the Functional Hardware Medium, Medium Verbs Provider, and Large stages.

**Comment by phender** [comment]
> The https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15409/13 build failed to run the Functional Hardware stages due to provisioning errors. I've manually kicked off https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15409/14/ to just run the Functional Hardware Medium, Medium Verbs Provider, and Large stages.

Functional HW testing passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15409/16/

**Review by tanabarr — APPROVED** [review]
looks like mostly clean cherry pick, although it's not a familiar area of code LGTM

**Pull request merged**